src/messages.rs

```diff
@@ -274,20 +293,81 @@ impl SignedMessage {
     /// Requires the list `src_sections` of nodes who should sign this message.
     pub fn new(content: RoutingMessage,
                full_id: &FullId,
-               mut src_sections: Vec<SectionList>)
+               mut src_sections: Vec<SignedSectionList>)
                -> Result<SignedMessage, RoutingError> {
-        src_sections.sort_by_key(|list| list.prefix);
+        src_sections.sort_by_key(|list| list.list.prefix);
         let sig = sign::sign_detached(&serialise(&content)?, full_id.signing_private_key());
         Ok(SignedMessage {
             content: content,
             src_sections: src_sections,
+            relay_sections: vec![],
             signatures: iter::once((*full_id.public_id(), sig)).collect(),
         })
     }
 
+    /// Add list of members of a relaying section
+    pub fn add_relaying_section(&mut self, section: SignedSectionList) {
+        self.relay_sections.push(section);
+    }
+
     /// Confirms the signatures.
-    // TODO (1677): verify the sending SectionLists via each hop's signed lists
-    pub fn check_integrity(&self, min_section_size: usize) -> Result<(), RoutingError> {
+    ///
+    /// `min_section_size` is required when the message is from a _group_ (see doc on `Authority`).
+    ///
+    /// `known_section` is used to kick-start hop-by-hop verification. It is required to verify
+    /// the source section/group; it should be mandatory but currently clients don't know who
+    /// the members of their proxy's section are (TODO).
+    pub fn check_integrity(&self,
+                           min_section_size: usize,
+                           known_section: Option<&SectionList>)
+                           -> Result<(), RoutingError> {
+        fn check_enough_sigs(ssl: &SignedSectionList,
+                             last_verified: &SectionList)
+                             -> Result<(), RoutingError> {
+            let bytes = serialise(&ssl.list)?;
+            // Now find all signatures of bytes from ssl which are valid and from a node
+            // mentioned by last_verified, and check that this is a quorum of last_verified.
+            let valid_sigs = ssl.signatures
+                .iter()
+                .filter(|&(pub_id, sig): &(&PublicId, &sign::Signature)| {
+                    last_verified.pub_ids.contains(pub_id) &&
+                    sign::verify_detached(sig, &bytes, pub_id.signing_public_key())
+                })
+                .count();
+            if QUORUM * last_verified.pub_ids.len() > 100 * valid_sigs {
+                return Err(RoutingError::NotEnoughSignatures);
+            }
+            Ok(())
+        }
+
+        // -----  first, verify the route  -----
+        if let Some(known_section) = known_section {
+            let mut last_verified = known_section;
+            for prev_hop in self.relay_sections.iter().rev() {
```

Comment:
Yes. We check each against the last verified section signature, then check the next against that... so reverse order is required.